### ISSUE TITLE
Opening with Chrome App Shortcuts causing infinite loop / not working

### ISSUE THREAD

**Issue by skoshy** [johnste/finicky#153]
If I try specifying a direct path to a Chrome App Shortcut (generated through Chrome Menu -> More Tools -> Create application shortcut), it causes an infinite loop of opening the window and doesn't actually go to the correct URL within the app as well.

Google Meet is a good example; if you have a Chrome App of https://meet.google.com and try clicking a Google Meet link from somewhere else, this behavior occurs.

Here's my config:

```js
const googleMeetApp = {
  browser: {
    name: '/Users/xxxxxx/Applications/Chrome Apps.localized/Meet.app',
    appType: 'appPath',
  },
}

module.exports = {
  defaultBrowser: "Google Chrome",
  rewrite: [
    {
      // Redirect all urls to use https
      match: ({ url }) => url.protocol === "http",
      url: { protocol: "https" }
    }
  ],
  handlers: [
    {
      match: finicky.matchHostnames(['meet.google.com']),
      ...googleMeetApp,
    },
    // ...
  ]
};

```

**Comment by johnste** [comment]
Hm not seeing the "Create application shortcut" menu item for google meet. There is a "Create shortcut" option but all it seems to do is to open up a normal tab in Chrome.

I did find an "install" option when I tried https://excalidraw.com/ which I might be able to use. It installs an app (does not open a browser window). I will try using that first.

**Comment by skoshy** [comment]
No problem @johnste , this issue is actually solved with PR https://github.com/johnste/finicky/pull/160 that I submitted a while ago.

Here's how I created an app shortcut for Google Meet:

![image](https://user-images.githubusercontent.com/369825/106373207-3c454d00-6345-11eb-8730-5c9eb4c592c6.png)

I then opened the Google Meet.app package and checked the Info.plist inside to get the app ID (mine was jhgifngagadhdmngcahhimadjfofillb).

I then set up Finicky like so:

```
const googleMeetApp = {
  browser: ({ urlString }) => ({
    name: "Google Chrome",
    profile: "Profile 13", // use whatever profile created the application shortcut
    args: [
      `--app-id=jhgifngagadhdmngcahhimadjfofillb`, // my app ID for my Google Meet application shortcut
      `--app-launch-url-for-shortcuts-menu-item=${urlString}`,
      // notice I'm not passing urlString as an array entry, since Chrome Application Shortcuts don't like that
    ],
  }),
}

module.exports = {
  // ...
  handlers: [
    {
      match: finicky.matchHostnames(["meet.google.com"]),
      ...googleMeetApp,
    },
  ],
};
```

This worked perfectly for opening Google Meet links in the newly made Google Meet app!

With this solution, I think this issue can be closed.

**Comment by johnste** [comment]
Great to hear! I'm closing this issue, and I added a link to it [in the wiki](https://github.com/johnste/finicky/wiki/Troubleshooting-&-Tips) in case it might be useful for someone else.

**Comment by schiegl** [comment]
The solution that @skoshy suggested does not work for app shortcuts created with Brave. The correct app opens, but only for the url the app shortcut was created for, not the custom url that was passed in `urlString`. Can anyone suggest a fix?

**Comment by skoshy** [comment]
Hi @schiegl, I actually switched to using Brave (beta) too and I can confirm it does work with Brave. Here's what my config file looks like:

```js
const workBrowser = {
  name: "com.brave.browser.beta",
  profile: "Profile 1"
};

const googleMeetApp = {
  browser: ({ urlString }) => ({
    ...workBrowser,
    args: [
      '--app-id=dbpkegghbgplfeaapbejbnfpbgnbjedf',
      `--app-launch-url-for-shortcuts-menu-item=${ urlString }`,
    ],
  }),
};

module.exports = {
  defaultBrowser: "com.brave.browser.beta",
  handlers: [
    {
      match: finicky.matchHostnames(["meet.google.com"]),
      ...googleMeetApp,
    },
    {
      // Open any link clicked in Slack in my work browser
      match: ({ opener }) =>
        opener.bundleId === "com.tinyspeck.slackmacgap",
      browser: workBrowser
    },
  ]
}
```

This config works great for me for Google Meet; we have a pinned Meet link to a specific meeting in our Slack group, and clicking it opens that exact meeting in the Google Meet Brave app. Other Google Meet links open appropriately as well.

**Comment by schiegl** [comment]
What a coincidence that you switched to Brave @skoshy! Using "Profile 1" seems to work for Google Meet, but not for Jira. Jira opens in a new tab instead of the app shortcut window (which also opens but remains blank). Also, the profile the shortcut was created in is not named "Profile 1". Not sure why "Profile 1" works?

Lastly, for some reason opening links through finicky takes much longer than opening shortcut apps with the system launcher. Instead of ~1s, it takes 5s. This is not due to page loading, as opening the webpage in a tab also just takes ~1s. Does this also happen to you?

**Comment by skoshy** [comment]
> Also, the profile the shortcut was created in is not named "Profile 1". Not sure why "Profile 1" works?

So the profile doesn't refer to the profile name, it refers to the profile folder. For me, my work profile in Brave is named `Work`, but the folder is actually `Profile 1`.

You can find all the profile directories in ~/Library/Application Support/BraveSoftware/Brave-Browser-Beta (or Brave-Browser).

The default profile is just called `Default`. There's also the `Guest Profile`. Any other profiles you make are numbered (`Profile 1`, `Profile 2`, etc).

> ...but not for Jira. Jira opens in a new tab instead of the app shortcut window (which also opens but remains blank)

Hmm, only advice for this is to check if the right profile directory is being used, and the correct `app-id`. I haven't tried this with JIRA, so can't offer any more advice than that :/

**Comment by schiegl** [comment]
> Hmm, only advice for this is to check if the right profile directory is being used, and the correct app-id. I haven't tried this with JIRA, so can't offer any more advice than that :/

The app-id is correct as the correct app shortcut opens. It is however blank. Given that the other apps work, maybe I matched the url wrong.

> For me, it adds maybe 1s delay, definitely not 5s though. That's an inherent problem with Finicky though I believe.

That is unfortunate.

**Comment by bigethan** [comment]
A short note, in chrome (with multiple profiles), specifying the profile caused this not to work (it'd open a blank chrome tab). Either commenting out that line( `// profile: ...` or setting it to `Default` made it work

**Comment by pauloermoraes** [comment]
Hello guys, if the shortcut app is already open, it opens another instance of the shortcut app for you too?
How can I open the URL on the same instance?

**Comment by baileyghjkl** [comment]
Hi @skoshy! I'm having the same issue as @bigethan; however, what I'm noticing is that regardless of what I do with the `profile: ...` line ("Profile Name" or literally any string, `Default`, or commenting out) it will not launch the app unless the last chrome window I interacted with corresponds to the same profile that created the app. Otherwise, it will open a new window/blank tab in the last active profile. Adding the account from the app-creator profile to the latest-interaction profile also has no effect on this behavior.

Any ideas on getting around this? I have to have 4 different profiles for my job(s) plus my personal account, and having to interact with the correct profile before a link will open in the right place is essentially the reason I found/installed finicky.  

Also, seconding @pauloermoraes's question about opening urls in the same instance (tho full disclosure I haven't done any digging on that yet to see if it's answered elsewhere).
